### DISCUSSION TITLE
Initialization of fields and properties with non-static values

### DISCUSSION THREAD

**Discussion by EldHasp** [dotnet/csharplang#5813] [Language Ideas]
Initialization of fields and properties with non-static values is possible only from the constructor.
Although at the time of their initialization, the memory area for the instance is allocated, which means this already has a value.
A typical example is when a command needs to be given a reference to an instance method. Both already exist at the time of initialization, but cannot be used.
You have to connect either a constructor or complicate the logic of obtaining a value by a property.

Instead of this initialization:
```
    public class ViewModel
    {
        public Command Command { get; }

        private void Execute()
        {
            // Some code
        }

        public ViewModel()
        {
            Command = new Command(Execute);
        }
    }
```

It is possible to such an initialization:
```
public class ViewModel
{
    public Command Command { get; } = new Command(Execute);

    private void Execute()
    {
        // Some code
    }
}
```

**Comment by HaloFour** [comment]
In C# field initializers are executed prior to the base constructor being called and that makes it illegal to reference `this` or any instance members due to not having been properly initialized.

**Reply by EldHasp** [reply]
I know it.
But! To initialize the fields need already allocated memory for the instance, which means that `this` can be materialized as soon as this memory is allocated.
Now this is done only when the base constructor is called, but nothing prevents it from being done immediately when memory is allocated for the instance.

**Reply by HaloFour** [reply]
Yes, the memory is allocated, but it is not initialized.  It's explicitly part of the design to try to prevent access to that memory until the base constructor is called.  Eliminating that rule without appropriate guardrails in place would definitely result in bugs that would be challenging to figure out.

**Reply by EldHasp** [reply]
I agree with you.
It will be the same as for static members now: sometimes the order in which they are written in the code is important, since they are initialized in the same order.
And sometimes have to initialize them from a static constructor to avoid uncertainty.
And in this respect, the initialization of instance fields is no different.
For her, the order of their declaration in the code will also be important, and in case of uncertainty they will be initialized in the constructor.
But in practice this will happen extremely rarely, just as this happens extremely rarely for static fields.
In addition, you can introduce a rule in which fields are first initialized without `this`, and then with them - although I think this is already redundant.

**Reply by masonwheeler** [reply]
> Eliminating that rule without appropriate guardrails in place would definitely result in bugs that would be challenging to figure out.

I don't agree.  It would result in the kind of bugs that would be *very obvious and easy to figure out* the first time you test the code.  (And if you *don't* test your newly-written code before deploying it, that's hardly the language's fault now is it?)

**Reply by HaloFour** [reply]
C# has made a number of opinionated design decisions specifically to avoid known sources of bugs.  Saying that those bugs would only happen to lazy developers is not productive nor does it lend much to the argument.

**Reply by CyrusNajmabadi** [reply]
@HaloFour is correct. This is the premise of pretty much all the rules in the language.  Finding and preventing these sorts of issues at compile time is essentially the premise of static languages.

**Comment by jnm2** [comment]
#140 is being implemented. When it ships, you will be able to do:

```cs
public Command Command => field ??= new(Execute);
```

Which will be the equivalent of today's:
```cs
private Command command;
public Command Command => command ??= new(Execute);
```

**Reply by EldHasp** [reply]
Yes, this will partially solve the problem.
But it is better to go further and remove this restriction altogether.
Nothing prevents materializing `this` as soon as memory is allocated for the instance, yet before the instance fields are initialized.
After all, the same thing happens with the initialization of static fields.
Memory is allocated for the object (type) and you can immediately refer to static fields. This does not have to be done from a static constructor.
Do the same for instance fields.

**Reply by DaZombieKiller** [reply]
The `Command` constructor could invoke the delegate given to it, which would cause `Execute` to run before the class is initialized. `Command` doesn't do this, but any other class could -- that's the problem that is solved by preventing this.

**Reply by jnm2** [reply]
> But it is better to go further and remove this restriction altogether.

@EldHasp Can you make a case for why this is better? Being technically possible, and being more like static fields, aren't necessarily practical enough to motivate changing the language spec and implementation. (Or losing the safety that @DaZombieKiller mentions.)

**Reply by EldHasp** [reply]
> The Command constructor could invoke the delegate given to it, which would cause Execute to run before the class is initialized.

Yes, this is possible.
But for static fields and a static constructor, this does not create problems.
So why should the same create problems for the instance constructor?

> Can you make a case for why this is better? 

Taking your example, then: `public Command Command { get; } = new(Execute);`.

**Reply by DaZombieKiller** [reply]
> So why should the same create problems for the instance constructor?

To rephrase, this is basically asking "if you can screw up and write dangerous code there, why can't we do it here too?", which answers the question itself I think :smile:. It would be a bad idea to introduce potentially dangerous behavior just because it allows for some terser syntax.

**Reply by jnm2** [reply]
@EldHasp That's not a practical example though because `field ??=` is being added to the language sooner, and it can be used here. What are motivating examples where `field ??=` would not help?

**Reply by EldHasp** [reply]
> What are motivating examples where field ??= would not help?

And what are the motivating arguments for initializing static fields?
After all, except for `??=` and in the constructor, they can be initialized immediately upon declaration.
This is exactly the same motivation for instance members.
Static fields can be initialized with values from other static fields, respectively, and instance fields must be able to be initialized with the values of instance members.
This will be understandable and the same type of behavior, and not as it is now - it is not clear where the tabu came from. 

> It would be a bad idea to introduce potentially dangerous behavior

Firstly, static fields already have such "potentially dangerous" behavior, so there can be no talk of a "new danger". It's just an extension of the static field rules to member fields.

Secondly, I don't see any potential danger here.
This is the same as confusing the order of initialization, assignment of values, calculation of local variables.
Is there such a "danger" leading to errors?
Of course there is.
But these are "programming routines": if you confused "+" with "-", then there is no one to blame but yourself for this.

**Reply by HaloFour** [reply]
IMO you're demonstrating exactly why the design decision here should not be reconsidered.  Slightly inconveniencing people who ignore the available workarounds is not good enough justification for reenabling an entire class of bugs.

**Reply by jnm2** [reply]
> And what are the motivating arguments for initializing static fields?

@EldHasp I didn't see you state a motivating argument for initializing static fields. "They can be initialized immediately upon declaration" is a description of the feature, not a motivation for it. What specifically needs the motivating argument is the change to the language; at this point, static fields don't need a motivation to be the way they already are. If they did have a motivation that also holds for instance fields (given instance fields' unique set of characteristics and usage patterns), that would be good to bring up.
